src/Formidable.js

```diff
@@ -538,17 +547,34 @@ class IncomingForm extends EventEmitter {
     return name;
   }
 
-  _setUpRename() {
-    const hasRename = typeof this.options.filename === 'function';
+  _joinDirectoryName(name) {
+    const newPath = path.join(this.uploadDir, name);
+
+    // prevent directory traversal attacks
+    if (!newPath.startsWith(this.uploadDir)) {
+      return path.join(this.uploadDir, this.options.defaultInvalidName);
+    }
 
-    if (this.options.keepExtensions === true && hasRename) {
-      this._rename = (part) => {
-        const resultFilepath = this.options.filename.call(this, part, this);
+    return newPath;
+  }
 
-        return this._uploadPath(part, resultFilepath);
+  _setUpRename() {
+    const hasRename = typeof this.options.filename === 'function';
+    if (hasRename) {
+      this._getNewName = (part) => {
+        let ext = '';
+        let name = this.options.defaultInvalidName;
+        if (part.filename) {
```

Comment:
should be more safe to be `!part.mime` as in `_handlePart`

Reply:
this._getNewName is only called when we already know it is a file. we have this if because sometimes filename can be null if it could not be parsed